kousa/lib/onion/chat.ex

```diff
@@ -177,6 +147,62 @@ defmodule Onion.RoomChat do
     {:noreply, %State{state | ban_map: Map.delete(state.ban_map, user_id)}}
   end
 
+  @spec send_msg(Kousa.Utils.UUID.t(), Broth.Message.Chat.Send.t()) :: :ok
+  def send_msg(room_id, payload) do
+    cast(room_id, {:send_msg, payload})
+  end
+
+  defp send_msg_impl(payload = %{from: from}, state) do
+    # throttle sender
+    with false <- should_throttle(from, state),
+         false <- user_banned?(from, state) do
+      dispatch_message(payload, state)
+      updated_message_map = Map.put(state.last_message_map, from, DateTime.utc_now())
+      new_state = %{state | last_message_map: updated_message_map}
+      {:noreply, new_state}
+    else
+      _ -> {:noreply, state}
+    end
+  end
+
+  @message_time_limit_milliseconds 1000
+  defp should_throttle(user, %{last_message_times: m})
+       when is_map_key(m, user) do
+    DateTime.diff(m[user], DateTime.utc_now(), :millisecond) >= @message_time_limit_milliseconds
+  end
+
+  defp should_throttle(_, _), do: false
+
+  defp dispatch_message(payload, state) do
+    case payload.whisperedTo do
+      [] ->
+        PubSub.broadcast("chat:" <> state.room_id, %Broth.Message{
+          operator: "chat:send",
+          payload: payload
+        })
+
```

Comment:
this is where pubsub messages are broadcast out to all of the listening websocket processes.